zkvm/src/ops.rs

```diff
@@ -399,4 +407,59 @@ impl Program {
         self.0.push(Instruction::Cloak(m, n));
         self
     }
+
+    /// Takes predicate and closure to add choose operations for
+    /// predicate tree traversal.
+    pub fn choose_predicate<F>(
+        &mut self,
+        pred: Predicate,
+        choose_fn: F,
+    ) -> Result<&mut Program, VMError>
+    where
+        F: FnOnce(PredicateTree) -> Result<&mut Program, VMError>,
```

Comment:
asking the closure to return `&mut Program` won't work well if it ends with left/right - which return Self. Since we have &mut Program as &mut self already, we can simply return it ourselves, allowing the closure to return anything (which would be `PredicateTree` or `()`, depending on the user's operations):

```rust
pub fn choose_predicate<F,T>(...) where
        F: FnOnce(PredicateTree) -> Result<T, VMError>
```